File: tauri/src-tauri/src/core/prompt/tool_format.rs
use super::PromptPart;
use crate::core::tools::{Tool, ToolCategory};

pub struct ToolFormatPart<'a> {
    pub tools: &'a [Box<dyn Tool>],
}

impl<'a> PromptPart for ToolFormatPart<'a> {
    fn render(&self) -> String {
        let categories = [
            (ToolCategory::FileSystem, "FILE SYSTEM"),
            (ToolCategory::CodeIntelligence, "CODE INTELLIGENCE"),
            (ToolCategory::Execution, "EXECUTION"),
            (ToolCategory::Communication, "COMMUNICATION"),
            (ToolCategory::General, "GENERAL"),
        ];

        let mut tool_sections = String::new();

        for (cat, cat_name) in &categories {
            let cat_tools: Vec<&Box<dyn Tool>> = self.tools.iter().filter(|t| &t.category() == cat).collect();
            if cat_tools.is_empty() { continue; }

            tool_sections.push_str(&format!("\n### {}\n\n", cat_name));

            for tool in cat_tools {
                let destructive_badge = if tool.is_destructive() { " ⚠️ DESTRUCTIVE" } else { "" };
                tool_sections.push_str(&format!("**{}**{}\n", tool.name(), destructive_badge));
                tool_sections.push_str(&format!("{}\n", tool.description()));

                // Parameters
                if let Ok(params) = serde_json::from_str::<serde_json::Value>(&tool.parameters()) {
                    if let Some(props) = params.get("properties").and_then(|p| p.as_object()) {
                        let required: Vec<String> = params.get("required")
                            .and_then(|r| r.as_array())
                            .map(|arr| arr.iter().filter_map(|v| v.as_str().map(String::from)).collect())
                            .unwrap_or_default();

                        tool_sections.push_str("Parameters:\n");
                        for (name, schema) in props {
                            let desc = schema.get("description").and_then(|d| d.as_str()).unwrap_or("");
                            let req = if required.contains(name) { " (required)" } else { " (optional)" };
                            tool_sections.push_str(&format!("  - `{}`{}: {}\n", name, req, desc));
                        }
                    }
                }

                // Examples
                let examples = tool.examples();
                if !examples.is_empty() {
                    for ex in &examples {
                        tool_sections.push_str(&format!("Example — {}:\n{}\n", ex.description, ex.xml));
                    }
                }

                tool_sections.push('\n');
            }
        }

        format!(
            "## AVAILABLE TOOLS\n\
             {}\n\
             ## TOOL CALLING FORMAT\n\n\
             To call a tool, use this XML format. All parameter values MUST be strings:\n\
             <tool_call>\n  <name>tool_name</name>\n  <parameters>\n    <param_name>value</param_name>\n  </parameters>\n</tool_call>\n\n\
             You can call ONE tool at a time. After each call, the system provides the result in a <tool_result> block.\n\
             Wait for the result before proceeding to the next step.",
            tool_sections
        )
    }
}

File: electron/electron/core/prompt/tool_format.ts
import type { PromptPart } from './types'
import type { Tool } from '../tools/types'
import { ToolCategory } from '../tools/types'

class ToolFormatPart implements PromptPart {
    private tools: Tool[]

    constructor(tools: Tool[]) {
        this.tools = tools
    }

    render(): string {
        const categories: [ToolCategory, string][] = [
            [ToolCategory.FileSystem, 'FILE SYSTEM'],
            [ToolCategory.CodeIntelligence, 'CODE INTELLIGENCE'],
            [ToolCategory.Execution, 'EXECUTION'],
            [ToolCategory.Communication, 'COMMUNICATION'],
            [ToolCategory.General, 'GENERAL'],
        ]

        let toolSections = ''

        for (const [cat, catName] of categories) {
            const catTools = this.tools.filter(t => t.category() === cat)
            if (catTools.length === 0) continue

            toolSections += `\n### ${catName}\n\n`

            for (const tool of catTools) {
                const badge = tool.isDestructive() ? ' ⚠️ DESTRUCTIVE' : ''
                toolSections += `**${tool.name()}**${badge}\n`
                toolSections += `${tool.description()}\n`

                try {
                    const params = JSON.parse(tool.parameters())
                    const props = params.properties || {}
                    const required: string[] = params.required || []

                    if (Object.keys(props).length > 0) {
                        toolSections += 'Parameters:\n'
                        for (const [name, schema] of Object.entries(props) as [string, any][]) {
                            const req = required.includes(name) ? ' (required)' : ' (optional)'
                            const desc = schema.description || ''
                            toolSections += `  - \`${name}\`${req}: ${desc}\n`
                        }
                    }
                } catch { /* Ignore malformed params */ }

                const examples = tool.examples()
                for (const ex of examples) {
                    toolSections += `Example — ${ex.description}:\n${ex.xml}\n`
                }

                toolSections += '\n'
            }
        }

        return `## AVAILABLE TOOLS
${toolSections}
## TOOL CALLING FORMAT

To call a tool, use this XML format. All parameter values MUST be strings:
<tool_call>
  <name>tool_name</name>
  <parameters>
    <param_name>value</param_name>
  </parameters>
</tool_call>

You can call ONE tool at a time. After each call, the system provides the result in a <tool_result> block.
Wait for the result before proceeding to the next step.`
    }
}

export default ToolFormatPart
